### PR TITLE
fix(vast): Validate VAST according to VAST2.0.

### PR DESCRIPTION
**Pull request by DevKai**
# What
This PR aims to make vungle/vungo fully support VAST2.0
# Acceptance
VAST2.0 should be correctly implemented.

### CODE REVIEW

File: vast/inline.go [hunk 1]
```diff
@@ -4,15 +4,15 @@ package vast
...
+	AdTitle     string        `xml:"AdTitle"`            // Required.
+	Impressions []*Impression `xml:"Impression"`         // Required.
+	Creatives   []*Creative   `xml:"Creatives>Creative"` // Creatives node is required.
 	Description string        `xml:"Description,omitempty"`
-	Advertiser  string        `xml:"Advertiser,omitempty"`
+	Advertiser  string        `xml:"Advertiser,omitempty"` // VAST3.0.
 	SurveyUrl   string        `xml:"Survey,omitempty"`
-	Errors      []string      `xml:"Error,omitempty"`
-	Pricing     *Pricing      `xml:"Pricing,omitempty"`
+	Errors      []string      `xml:"Error,omitempty"`   // Error should be an URI in VAST2.0.
```
Comment (garukun) [inline vast/inline.go]: an -> a

Comment (garukun) [inline vast/inline.go]: Also in addition can u add this to the validate method?

File: vast/linear.go [hunk 2]
```diff
@@ -4,15 +4,15 @@ import "github.com/Vungle/vungo/vast/defaults"
 
 // Linear type represents an <Linear> element within a <InLine> element.
 type Linear struct {
-	SkipOffset *Offset `xml:"skipoffset,attr,omitempty"` // Time delay at which ad becomes skippable; if absent, the ad is unskippable.
+	SkipOffset *Offset `xml:"skipoffset,attr,omitempty"` // Time delay at which ad becomes skippable; if absent, the ad is unskippable. VAST3.0.
```
Comment (garukun) [inline vast/linear.go]: This in line comment is a bit long please change to godoc

File: vast/linearwrapper.go [hunk 3]
```diff
@@ -3,10 +3,13 @@ package vast
...
 // wrapped linear creative.
 type LinearWrapper struct {
-	Icons       []*Icon      `xml:"Icons>Icon,omitempty"`
+	Icons       []*Icon      `xml:"Icons>Icon,omitempty"` // VAST3.0.
 	Trackings   []*Tracking  `xml:"TrackingEvents>Tracking,omitempty"`
-	VideoClicks *VideoClicks `xml:"VideoClicks,omitempty"`
-	Extensions  []*Extension `xml:"CreativeExtensions>CreativeExtension,omitempty"`
+	VideoClicks *VideoClicks `xml:"VideoClicks,omitempty"` // VideoClicks here is different from the Linear one.
+	// VideoClicks should be a slice of ClickTracking in VAST2.0. This line should be like :
+	// VideoClicks []*VideoClick `xml:"VideoClicks>ClickTracking,omitempty"`
```
Comment (garukun) [inline vast/linearwrapper.go]: So fix it?

File: vast/nonlinearwrapper.go [hunk 4]
```diff
@@ -4,18 +4,25 @@ package vast
...
 	ExpandedHeight            int       `xml:"expandedHeight,attr"`
 	IsScalable                bool      `xml:"scalable,attr,omitempty"`
 	ShouldMaintainAspectRatio bool      `xml:"maintainAspectRatio,attr,omitempty"`
 	MinSuggestedDuration      *Duration `xml:"minSuggestedDuration,attr,omitempty"`
 	ApiFramework              string    `xml:"apiFramework,attr,omitempty"`
 
-	Trackings      []*Tracking  `xml:"TrackingEvents>Tracking,omitempty"`
-	ClickTrackings []string     `xml:"NonLinearClickTracking,omitempty"`
-	Extensions     []*Extension `xml:"CreativeExtensions>CreativeExtension,omitempty"`
+	// Missing fields in VAST2.0.
```
Comment (garukun) [inline vast/nonlinearwrapper.go]: Are we current reading these fields?

File: vast/vast.go [hunk 5]
```diff
@@ -5,9 +5,9 @@ import "encoding/xml"
 // Vast type represents the root <VAST> tag.
 type Vast struct {
 	XMLName xml.Name `xml:"VAST"`
-	Version Version  `xml:"version,attr"`
+	Version Version  `xml:"version,attr"` // required.
```
Comment (garukun) [inline vast/vast.go]: required -> Required

File: vast/videoclicks.go [hunk 6]
```diff
@@ -2,7 +2,7 @@ package vast
 
 // VideoClicks type represents a <VideoClicks> element that contains various types of VideoClick's.
 type VideoClicks struct {
-	ClickThroughs  []*VideoClick `xml:"ClickThrough,omitempty"`
+	ClickThroughs  []*VideoClick `xml:"ClickThrough,omitempty"` // ClickThrough should be an VideoClick not a slice in VAST2.0.
```
Comment (garukun) [inline vast/videoclicks.go]: I believe this will still work? An array of one

File: vast/wrapper.go [hunk 7]
```diff
@@ -5,11 +5,11 @@ package vast
...
 type Wrapper struct {
-	AdSystem     *AdSystem          `xml:"AdSystem"`
-	VastAdTagUri string             `xml:"VASTAdTagURI"`
-	Impressions  []*Impression      `xml:"Impression"`
-	Errors       []string           `xml:"Error,omitempty"`
-	Creatives    []*CreativeWrapper `xml:"Creatives>Creative"`
+	AdSystem     *AdSystem          `xml:"AdSystem"`           // Required.
+	VastAdTagUri string             `xml:"VASTAdTagURI"`       // Required.
+	Impressions  []*Impression      `xml:"Impression"`         // Required. should be a URI slice in VAST2.0.
+	Errors       []string           `xml:"Error,omitempty"`    // Error should be an URI instead of a slice in VAST2.0.
```
Comment (garukun) [inline vast/wrapper.go]: Still works? Array of one?

File: vast/adparameters.go [hunk 8]
```diff
@@ -4,15 +4,14 @@ package vast
...
+	IsXmlEncoded bool `xml:"xmlEncoded,attr,omitempty"` // VAST3.0.
 }
 
 // Validate method validates the AdParameters according to the VAST.
 // Parameters are required.
 func (p *AdParameters) Validate() error {
-
 	if len(p.Parameters) == 0 {
-		return ErrAdParametersMissParameters
+		return ValidationError{Errs: []error{ErrAdParametersMissParameters}}
```
Comment (WeiVungle) [inline vast/adparameters.go]: Please be consistent.

File: vast/adsystem.go [hunk 9]
```diff
@@ -11,7 +11,7 @@ type AdSystem struct {
 // System is required.
 func (as *AdSystem) Validate() error {
 	if len(as.System) == 0 {
-		return ErrAdSystemMissSystem
+		return ValidationError{Errs: []error{ErrAdSystemMissSystem}}
```
Comment (WeiVungle) [inline vast/adsystem.go]: As above.

File: vast/delivery.go [hunk 10]
```diff
@@ -12,7 +12,7 @@ const (
 // Validate method validates Delivery according to the VAST.
 func (delivery Delivery) Validate() error {
 	if delivery != DELIVERY_PROGRESSIVE && delivery != DELIVERY_STREAMING {
-		return ErrUnsupportedDeliveryType
+		return ValidationError{Errs: []error{ErrUnsupportedDeliveryType}}
```
Comment (WeiVungle) [inline vast/delivery.go]: Please be consistent.

File: vast/duration.go [hunk 11]
```diff
@@ -73,10 +73,10 @@ func (d Duration) String() string {
 // The Duration should never less than zero.
 func (d Duration) Validate() error {
 	if d < 0 {
-		return ErrDurationNegative
+		return ValidationError{Errs: []error{ErrDurationNegative}}
```
Comment (WeiVungle) [inline vast/duration.go]: As above.

File: vast/duration.go [hunk 12]
```diff
@@ -73,10 +73,10 @@ func (d Duration) String() string {
...
 	}
 	if d == 0 {
-		return ErrDurationEqualZero
+		return ValidationError{Errs: []error{ErrDurationEqualZero}}
```
Comment (WeiVungle) [inline vast/duration.go]: As above.

File: vast/event.go [hunk 13]
```diff
@@ -48,7 +48,7 @@ func (e Event) Validate() error {
 	case EVENT_SKIP:
 	case EVENT_PROGRESS:
 	default:
-		return ErrUnsupportedEvent
+		return ValidationError{Errs: []error{ErrUnsupportedEvent}}
```
Comment (WeiVungle) [inline vast/event.go]: As above.

File: vast/htmlresource.go [hunk 14]
```diff
@@ -11,7 +11,7 @@ type HtmlResource struct {
 func (r *HtmlResource) Validate() error {
 
 	if len(r.Html) == 0 {
-		return ErrHtmlResourceMissHtml
+		return ValidationError{Errs: []error{ErrHtmlResourceMissHtml}}
```
Comment (WeiVungle) [inline vast/htmlresource.go]: As above.

File: vast/impression.go [hunk 15]
```diff
@@ -13,7 +13,7 @@ type Impression struct {
 func (impression *Impression) Validate() error {
 
 	if len(impression.Uri) == 0 {
-		return ErrImpressionMissUri
+		return ValidationError{Errs: []error{ErrImpressionMissUri}}
```
Comment (WeiVungle) [inline vast/impression.go]: As above.

File: vast/mode.go [hunk 16]
```diff
@@ -12,9 +12,8 @@ const (
 
 // Validate method validates Mode according to the VAST.
 func (mode Mode) Validate() error {
-
 	if mode != MODE_ALL && mode != MODE_ANY && mode != MODE_NONE {
-		return ErrUnsupportedMode
+		return ValidationError{Errs: []error{ErrUnsupportedMode}}
```
Comment (WeiVungle) [inline vast/mode.go]: As above.

File: vast/pricingmodel.go [hunk 17]
```diff
@@ -20,7 +20,7 @@ func (module PricingModel) Validate() error {
 	case PRICING_MODEL_CPE:
 	case PRICING_MODEL_CPV:
 	default:
-		return ErrUnsupportedPriceModelType
+		return ValidationError{Errs: []error{ErrUnsupportedPriceModelType}}
```
Comment (WeiVungle) [inline vast/pricingmodel.go]: As above.

File: vast/staticresource.go [hunk 18]
```diff
@@ -3,16 +3,15 @@ package vast
...
 	Uri      TrimmedData `xml:",cdata"`
 }
 
 // Validate method validates the StaticResource according to the VAST.
 // Uri is required.
 func (r *StaticResource) Validate() error {
-
 	if len(r.Uri) == 0 {
-		return ErrStaticResourceMissUri
+		return ValidationError{Errs: []error{ErrStaticResourceMissUri}}
```
Comment (WeiVungle) [inline vast/staticresource.go]: As above.

File: vast/version.go [hunk 19]
```diff
@@ -11,9 +11,8 @@ const (
 
 // Validate method validates the Version according to the VAST.
 func (v Version) Validate() error {
-
 	if v != VERSION_2 && v != VERSION_3 {
-		return ErrUnsupportedVersion
+		return ValidationError{Errs: []error{ErrUnsupportedVersion}}
```
Comment (WeiVungle) [inline vast/version.go]: As above.

**Comment by VG-HarrisLiu** [comment]
Logical is no problem,  but expect  in the further,  we can split vast2 and vast3 in different package. let the defines clearly.

File: vast/videoclicks.go [hunk 20]
```diff
@@ -10,28 +10,41 @@ type VideoClicks struct {
...
 func (vc *VideoClicks) Validate() error {
+	errors := make([]error, 0)
 
-	if len(vc.ClickThroughs) == 0 {
-		return ErrVideoClicksMissClickThroughs
-	}
+	// TODO(@DevKai): VAST3.0 require ClickThroughs not be empty.
+	//if len(vc.ClickThroughs) == 0 {
+	//	errors = append(errors, ErrVideoClicksMissClickThroughs)
+	//}
```
Comment (xiaopeng0216) [inline vast/videoclicks.go]: nitpick: please don't leave commented codes.

File: vast/adparameters.go (hunk 8 above)
Comment (DevKai) [inline vast/adparameters.go]: Since there is only one type of error, it would be concise.

**Comment by xiaopeng0216** [comment]
PTAL

**Comment by xiaopeng0216** [comment]
LGTM

**Comment by tiagozortea** [comment]
LGTM

**Pull request merged**